atomate/vasp/workflows/base/magnetism.py

```diff
@@ -0,0 +1,834 @@
+        symmetry arguments alone which orderings to prefer. Here, we use a
+        variety of strategies (heuristics) to enumerate plausible orderings,
+        and later discard any duplicates that might be found by multiple
+        strategies. This approach is not ideal, but has been found to be
+        relatively robust over a wide range of magnetic structures.
+        """
+
+        formula = structure.composition.reduced_formula
+        transformations = {}
+
+        # analyzer is used to obtain information on sanitized input
+        analyzer = CollinearMagneticStructureAnalyzer(
+            structure,
+            default_magmoms=self.default_magmoms,
+            overwrite_magmom_mode="replace_all",
+        )
+
+        if not analyzer.is_magnetic:
+            raise ValueError(
+                "Not detected as magnetic, add a new default magmom for the "
+                "element you believe may be magnetic?"
+            )
+
+        # now we can begin to generate our magnetic orderings
+        logger.info("Generating magnetic orderings for {}".format(formula))
+
+        mag_species_spin = analyzer.magnetic_species_and_magmoms
+        types_mag_species = sorted(
+            analyzer.types_of_magnetic_specie,
+            key=lambda sp: analyzer.default_magmoms.get(str(sp), 0),
+            reverse=True,
+        )
+        types_mag_elements = {sp.symbol for sp in types_mag_species}
+        num_mag_sites = analyzer.number_of_magnetic_sites
+        num_unique_sites = analyzer.number_of_unique_magnetic_sites()
+
+        # enumerations become too slow as number of unique sites (and thus
+        # permutations) increase, 8 is a soft limit, this can be increased
+        # but do so with care
+        if num_unique_sites > self.max_unique_sites:
+            raise ValueError("Too many magnetic sites to sensibly perform enumeration.")
+
+        # maximum cell size to consider: as a rule of thumb, if the primitive cell
+        # contains a large number of magnetic sites, perhaps we only need to enumerate
+        # within one cell, whereas on the other extreme if the primitive cell only
+        # contains a single magnetic site, we have to create larger supercells
+        if "max_cell_size" not in self.transformation_kwargs:
+            # TODO: change to 8 / num_mag_sites ?
+            self.transformation_kwargs["max_cell_size"] = max(1, int(4 / num_mag_sites))
+        logger.info(
+            "Max cell size set to {}".format(
+                self.transformation_kwargs["max_cell_size"]
+            )
+        )
+
+        # when enumerating ferrimagnetic structures, it's useful to detect
+        # symmetrically distinct magnetic sites, since different
+        # local environments can result in different magnetic order
+        # (e.g. inverse spinels)
+        # initially, this was done by co-ordination number, but is
+        # now done by a full symmetry analysis
+        sga = SpacegroupAnalyzer(structure)
```

Comment:
Some of this code regarding finding unique sites for magnetization - why put in atomate rather than pymatgen?

In general I see pymatgen as the place for doing any algorithm development, atomate should mostly be calling pymatgen with some settings for high-throughput workflows.

Reply:
especially since (as far as I can tell) your input and output objects are not workflows, fireworks etc, this is just structure enumeration

Reply:
Yes, it's definitely a balance. I also commit a lot of improvements to the `MagOrderingTransformation` in pymatgen and a new `MagOrderParameterConstraint` class too. The balance I ended up falling on was "transformation stuff goes into pymatgen, but *choice* of transformation kwargs goes into the workflow" -- and that is what this extra stuff is.

The design of this workflow is a little different in one respect though, there's the class itself, and then a `.get_wf()` method -- the reason for this choice was that I wanted more introspection into the workflow generation, and I also wanted to separate different components out for readability/ease of maintenance.